src/notebooks/controllers/types.ts

```diff
@@ -34,9 +34,13 @@ export const IControllerRegistration = Symbol('IControllerRegistration');
 
 export interface IControllerRegistration {
     /**
-     * Gets the list of all of the controllers
+     * Gets the filtered list of all of the controllers
      */
     values: IVSCodeNotebookController[];
+    /**
+     * Gets every registered connection metadata
+     */
+    all: KernelConnectionMetadata[];
```

Comment:
Can't we already get this same information from outside using the existing public property `values` as follows
`values.map(item => item.connection`

Adding another property seems redundant and complicates it as we have more than one way to get the same information. I.e. i though this would be different from `values`, however looking at the code this is identical.

I think we should remove the propert `registeredMetadatas` and make it just return as follows:
```typescript
private get metadatas() {
    return this.values.map(item => item.connection);
}
```

& then delete `this.regsiteredMetadatas` property.
I think i see a bug, instead of removing the controller when Uris change, we're only removing the metadata. Removing this duplicate property would resolve that bug

Reply:
values is not the same as 'all'. Values are the items active in the kernel picker. All is every kernel that's been registered.

When the URI changes we remove the metadatas and then we reapply the filter. This disposes all controllers that aren't allowed in the filter as shown here:
https://github.com/microsoft/vscode-jupyter/blob/4d1a5c7f58700fa93233226afddc061bd4d5878a/src/notebooks/controllers/controllerRegistration.ts#L245

Reply:
Oh ok, got it.